crates/ark/src/fixtures/dummy_frontend.rs

```diff
@@ -52,36 +53,39 @@ impl DummyArkFrontend {
             panic!("Can't spawn Ark more than once");
         }
 
-        let frontend = DummyFrontend::new();
-        let connection_file = frontend.get_connection_file();
-
         // We don't want cli to try and restore the cursor, it breaks our tests
         // by adding unecessary ANSI escapes. We don't need this in Positron because
         // cli also checks `isatty(stdout())`, which is false in Positron because
         // we redirect stdout.
         // https://github.com/r-lib/cli/blob/1220ed092c03e167ff0062e9839c81d7258a4600/R/onload.R#L33-L40
         unsafe { std::env::set_var("R_CLI_HIDE_CURSOR", "false") };
 
-        // Start the kernel in this thread so that panics are propagated
-        crate::start::start_kernel(
-            connection_file,
-            vec![
-                String::from("--interactive"),
-                String::from("--vanilla"),
-                String::from("--no-save"),
-                String::from("--no-restore"),
-            ],
-            None,
-            session_mode,
-            false,
-        );
-
-        // Start the REPL in a background thread, does not return and is never joined
-        stdext::spawn!("dummy_kernel", || {
+        let connection = DummyConnection::new();
+        let (connection_file, registration_file) = connection.get_connection_files();
+
+        // Start the kernel and REPL in a background thread, does not return and is never joined.
+        // Must run `start_kernel()` in a background thread because it blocks until it receives
+        // a `HandshakeReply`, which we send from `from_connection()` below.
+        stdext::spawn!("dummy_kernel", move || {
+            crate::start::start_kernel(
+                connection_file,
+                Some(registration_file),
+                vec![
+                    String::from("--interactive"),
+                    String::from("--vanilla"),
+                    String::from("--no-save"),
+                    String::from("--no-restore"),
+                ],
+                None,
+                session_mode,
+                false,
+            );
+
             RMain::start();
         });
```

Comment:
@lionel- since we are moving the kernel `setup()` back onto the same thread as the `RMain::start()`, I think we can move `R_MAIN_THREAD_ID` back into the `setup()` method to solve https://github.com/posit-dev/positron/issues/4973

In fact, after we merge this we can probably combine the `RMain::setup()` and `RMain::start()` again. I see no reason they should be separate now.